### PR TITLE
Add file content as a parameter for the tsserver open command

### LINKED ISSUE #5530: tsserver: need to be able to send content on open
Since the tsserver can crash (for example due to memory limits) we added code to VSCode to restart the server. However if a TS file is not saved we can't bring the server back in sync since the open command reads the content from disk. To enable the restart case the open command must accept the content as well. To not break existing client the content can be made optional and if not provided read from disk.

### PR DESCRIPTION
**Pull request by zhengbli**
Fix #5530.

### CODE REVIEW

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -1006,14 +1006,15 @@ namespace ts.server {
...
          */
-        openFile(fileName: string, openedByClient: boolean) {
+        openFile(fileName: string, openedByClient: boolean, fileContent?: string) {
             fileName = ts.normalizePath(fileName);
             let info = ts.lookUp(this.filenameToScriptInfo, fileName);
             if (!info) {
                 let content: string;
                 if (this.host.fileExists(fileName)) {
-                    content = this.host.readFile(fileName);
+                    content = fileContent ? fileContent : this.host.readFile(fileName);
```
Comment (sandersn) [inline src/server/editorServices.ts]: couldn't this be `fileContent || this.host.readFile(filename)` ?

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -1006,14 +1006,15 @@ namespace ts.server {
 
         /**
          * @param filename is absolute pathname
+         * @param fileContent is a known version of the file content that is more up to date
```
Comment (sandersn) [inline src/server/editorServices.ts]: More up to date than what? The content stored on the host, right? Can you update the comment with that detail?

Comment (zhengbli) [inline src/server/editorServices.ts]: Is "more up to date than the one on disk" making sense?
So the situation is that the server may have crashed before the client saves the edited buffer in sublime or vs code, therefore when restarting the server, the buffer content should take higher priority.

Comment (sandersn) [inline src/server/editorServices.ts]: Yes, that sounds good.

**Comment by zhengbli** [comment]
@mhegazy @sandersn any other comments?

File: src/server/protocol.d.ts [hunk 3]
```diff
@@ -513,6 +513,7 @@ declare namespace ts.server.protocol {
       *  Information found in an "open" request.
       */
     export interface OpenRequestArgs extends FileRequestArgs {
+        fileContent?: string;
```
Comment (mhegazy) [inline src/server/protocol.d.ts]: please add comment on how this is expected to work.

**Comment by mhegazy** [comment]
please add tests.

**Comment by mhegazy** [comment]
looks good. please add the comment and a test to ensure we do not break this in the future.

**Comment by sandersn** [comment]
Yes, looks good to me.

**Comment by dbaeumer** [comment]
@zhengbli: thanks for looking into this!
@mhegazy: to which branches will this be delivered. Any change to get this on 1.6.x which is still the version we are sitting on?

**Comment by mhegazy** [comment]
the best we can do is put it in 1.7. it is definitely going to be in 1.8.

**Comment by mhegazy** [comment]
is this blocking?

**Comment by zhengbli** [comment]
@mhegazy it might be blocking if it is used often with large files frequently I suppose, however the command is not supposed to be called very often (as suggested only in situations like when the server crashes)

**Comment by mhegazy** [comment]
:+1:

**Comment by mhegazy** [comment]
we can merge/cherry pick this into 1.7 as needed.

**Comment by mhegazy** [comment]
@dbaeumer would 1.7 work for you?

**Comment by dbaeumer** [comment]
@mhegazy we are still shipping 1.6.2. We did our testing last week and don't want to change the tsserver now. In addition the default TS npm package still seems to be 1.6.2. 
With the help of Alex we put a work around into our system which makes the behavior Andres as seen go away. But it doesn't solve the general problem. So given the fact that 1.7 will come out soon it is ok to only have it in 1.7.

**Comment by billti** [comment]
:+1:

**Pull request merged**